### PR TITLE
Add SFall arrays and patry_member_list opcodes

### PR DESCRIPTION
**Pull request by roginvs**
This PR adds support for SFall arrays. Opcodes:

+ CreateArray
+ SetArray
+ GetArray
+ FreeArray
+ LenArray
+ ResizeArray
+ TempArray
+ FixArray
+ StringSplit
+ ScanArray
+ Substr
+ TypeOf
+ GetArrayKey
+ StackArray
+ PartyMemberList

Tested using this script https://gist.github.com/phobos2077/6bf357c49caaf515371a13f9a2d74a41 (except loading and saving which are not implemented in this PR)

Need review regarding C++ related stuff

Related to https://github.com/alexbatalov/fallout2-ce/issues/29

### CODE REVIEW

**Comment by bademux** [comment]
> Not every opcode is tested because I do not have script where i can test opcodes implementation. Maybe we need to introduce unit testing framework here.

Sorry jumping into the thread, but it looks familiar to what I propose recently https://github.com/alexbatalov/fallout2-ce/issues/225
Implementation-wise, I see it as graalvm managed llvm bitcode as a test subject and JS\Groovy-Spock from scenario side.
More about Graalvm https://www.graalvm.org/latest/reference-manual/llvm/Interoperability/.
Let me know what you think.

**Comment by phobos2077** [comment]
I have a "unit-test"-like script for arrays that I've used when implementing some of those features originally, you can try it out:

https://gist.github.com/phobos2077/6bf357c49caaf515371a13f9a2d74a41

**Review by phobos2077 — COMMENTED** [review]
I was just thinking about porting this. You did an amazing work! Only saving arrays is left. Do you plan to implement it?

**Comment by roginvs** [comment]
Yes, I would like to implement missing items but after review or merge of this PR. Baby steps :)

**Comment by roginvs** [comment]
Changed state into "Draft" because:
  - Tests above are not passing
  - Strings are not copied into array elements and this might cause a problem
In summary: still working on it

**Comment by alexbatalov** [comment]
Hey, I've updated some of your code. Please review and if you're ok with this, let's move on.

File: src/sfall_arrays.cc [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+        return nullptr;
+    }
+
+    return it->second.get();
+}
+
+ProgramValue GetArrayKey(ArrayId arrayId, int index, Program* program)
+{
+    auto arr = get_array_by_id(arrayId);
+    if (arr == nullptr) {
```
Comment (roginvs) [inline src/sfall_arrays.cc:550]: As far as I know build-in pointer-to-bool conversion works exactly this way. Just a matter of style

File: src/sfall_arrays.cc [hunk 2]
```diff
@@ -0,0 +1,697 @@
...
+                return a.value < b.value;
+            });
+        } else {
+            ListSort(pairs, type, [](const KeyValuePair& a, const KeyValuePair& b) -> bool {
+                return a.key < b.key;
+            });
+        }
+    }
+
+    std::vector<KeyValuePair> pairs;
```
Comment (roginvs) [inline src/sfall_arrays.cc:472]: LOL no `map` for assoc arrays. I think it is fine if we know that arrays are usually small (for example less than 100 or 1000 items). Anyway I do not see any problem here

**Review by roginvs — COMMENTED** [review]
Hmm, ok, looks good for me!

PS: I prefer to throw if something unexpected happens, for example unexpected enum value inside switch statement. It is better to identify error earlier that later. On the other hand a smart compiler should warn if switch statement did not check all enum values, is not it?

**Comment by alexbatalov** [comment]
Adds many functions towards #200

**Comment by phobos2077** [comment]
I don't think the original Olympus uses any of advanced sfall features (you probably tested on of those "fixed" versions)? Anyway, great to see we are one step closer to sfall compatibility. Next step: global scripts.

**Pull request merged**